### ISSUE TITLE
Add color-function-notation

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#4718]
> What is the problem you're trying to solve?

I want to disallow the use of (comma-based) legacy color function notation so that I only wrote (space-based) future-facing syntax, and transpile it if necessary with [postcss-preset-env](https://preset-env.cssdb.org/features).

- [Spec](https://www.w3.org/TR/css-color-4/#rgb-functions)
- [CanIUse](https://caniuse.com/#feat=mdn-css_types_color_space_separated_functional_notation)
- [Relevant tweet](https://twitter.com/mathias/status/1253242715304857601)

Legacy notation:

```css
a { color: rgb(0, 0, 0, 0.5) }
```

Modern notation:

```css
a { color: rgb(0 0 0 / 50%) }
```

> What solution would you like to see?

A new rule to do this. Some users, like with pseudo-element colon notation, might want to enforce the legacy style. I propose a rule with two options: "space" and "comma" (unless there are better names?).

Any thoughts on the use case, rule name or options?

**Comment by evilebottnawi** [comment]
Sounds good! `color-function-notation: "space" | "comma"` - good name and options

**Comment by mathiasbynens** [comment]
👍 A stylelint option makes more sense than a Prettier option (which I had previously suggested [here](https://github.com/prettier/prettier/issues/8154)). Copy-pasting some ideas/opinions:

**Input:**
```css
foo {
  color: rgb(0, 128, 255);
  color: rgba(0, 128, 255, 0.5);
  color: hsl(198, 28%, 50%);
  color: hsla(198, 28%, 50%, 0.5);
  color: lab(56.29% 019.93 16.58 / 50%);
  color: lch(56.29% 19.86 236.62 / 50%);
  /* color(sRGB 0 0.50 1 / 50%); */
}
```

**Output:**
```css
foo {
  color: rgb(0, 128, 255);
  color: rgba(0, 128, 255, 0.5);
  color: hsl(198, 28%, 50%);
  color: hsla(198, 28%, 50%, 0.5);
  color: lab(56.29% 019.93 16.58 / 50%);
  color: lch(56.29% 19.86 236.62 / 50%);
  /* color(sRGB 0 0.50 1 / 50%); */
}
```

**Expected behavior:**
```css
foo {
  color: rgb(0 128 255);
  color: rgb(0 128 255 / 50%);
  color: hsl(198deg 28% 50%);
  color: hsl(198deg 28% 50% / 50%);
  color: lab(56.29% 019.93 16.58 / 50%);
  color: lch(56.29% 19.86 236.62deg / 50%);
  /* color(sRGB 0 0.50 1 / 50%); */
}
```

Notes:

- in general, [use the modern comma-free syntax instead of the legacy comma-based syntax](https://twitter.com/mathias/status/1253242715304857601) ("legacy" is how the CSS Color spec refers to it)
- use `rgb()` with the new syntax instead of `rgba()` (which is a legacy function)
- use `hsl()` with the new syntax instead of `hsla()` (which is a legacy function)
- consider being opinionated w.r.t. how to represent alpha values. IMHO using percentage notation is most readable, as it clarifies the range: e.g. `70%` vs `0.7`
- consider being opinionated about using the `deg` suffix where applicable, e.g. in the first parameter for `hsl()` and the last non-alpha parameter for `lch()`. Again, because this clarifies that the range lies between `0-360`.
- the `color()` syntax is not yet supported by the parser, but that's a separate issue.
- optionally, consider being opinionated about preferring `#rrggbba`/`#rgba` hex notation over `rgb()` notation, or the other way around

**Comment by jeddy3** [comment]
@mathiasbynens Thanks for chiming in. These colour conventions are an excellent fit for stylelint.

> - use `rgb()` with the new syntax instead of `rgba()` (which is a legacy function)
> - use `hsl()` with the new syntax instead of `hsla()` (which is a legacy function)

Thank you for highlighting these. The rule should address `rgba` and `hsla` functions as they are directly connected to comma(less) notation.

With the "commaless" option the following is invalid:

```css
a { color: rgba(0, 128, 255, 0.5); }
```

And the rule should autofix it to:

```css
a { color: rgb(0 128 255 / 0.5); }
```

Likewise for `hsla` functions.

> - consider being opinionated w.r.t. how to represent alpha values. IMHO using percentage notation is most readable, as it clarifies the range: e.g. `70%` vs `0.7`
> - consider being opinionated about using the `deg` suffix where applicable, e.g. in the first parameter for `hsl()` and the last non-alpha parameter for `lch()`. Again, because this clarifies that the range lies between `0-360`.

These sound good to me. I've created new issues for them: https://github.com/stylelint/stylelint/issues/4720 and https://github.com/stylelint/stylelint/issues/4721, respectively.

> - optionally, consider being opinionated about preferring `#rrggbba`/`#rgba` hex notation over `rgb()` notation, or the other way around

This can be done with a combination of the following rules:

- [color-no-hex](https://stylelint.io/user-guide/rules/)
- [color-hex-length](https://stylelint.io/user-guide/rules/color-hex-length)
- [color-named](https://stylelint.io/user-guide/rules/color-named)
- [function-blacklist](https://stylelint.io/user-guide/rules/function-blacklist)

However, it's not possible to autofix a preference for hex or `rgb` notation with this approach. The community [`stylelint-color-format`](https://www.npmjs.com/package/stylelint-color-format) plugin can do this, but I'll create an issue to discuss bringing this functionality into stylelint itself.

---

Fleshing out the rule using [selector-pseudo-element-colon-notation](https://stylelint.io/user-guide/rules/selector-pseudo-element-colon-notation), which also addresses legacy syntax:

- Name: `color-function-notation`
- Primary option: `"comma"|"commaless"`
- Secondary options: None
- Message: `Expected ${option} color-function notation`
- Description: "Specify comma or commaless notation for applicable color-functions."
- Section: ["Limit language features" -> "Color"](https://stylelint.io/user-guide/rules/list#color-1)

Does anyone have any more thoughts, or should l label this rule as good-to-go?

**Comment by mathiasbynens** [comment]
Everything you say makes a lot of sense to me, @jeddy3! One tiny suggestion: `commaless` sounds a bit awkward IMHO. Maybe `"comma"|"modern"`?

**Comment by jeddy3** [comment]
Yes, I agree.

> in general, use the _modern comma-free_ syntax instead of the _legacy comma-based_ syntax [emphasis mine]

On a second look, I prefer the "comma-free" and "comma-based" terms that you used in your original comment. They fit nicely with our use pronunciation terms in [the names](https://stylelint.io/user-guide/rules/about#whitespace-rules) and [options](https://stylelint.io/user-guide/rules/selector-pseudo-element-colon-notation#options) of the rules. We can use "modern" and "legacy" terms in the [extended description](https://stylelint.io/user-guide/rules/about#readmes) of the rules. We do this already for [selector-pseudo-element-colon-notation](https://stylelint.io/user-guide/rules/selector-pseudo-element-colon-notation), where we use the [words of the spec](https://www.w3.org/TR/css-pseudo-4/#css2-compat) in the extended description.

We can also flesh out the extended description with the rationale for the syntax [mentioned in this twitter thread](https://twitter.com/tabatkins/status/1253364375575883777) to help inform curious users.

Changes to rule blueprint:

- Primary option: `"comma-free"|"comma-based"`
- Description: "Specify comma-free or comma-based notation for applicable color-functions."
- Extended description: 

   "Modern color-functions use a comma-free syntax because functions in CSS are used to group/name a syntax chunk. They should work by the same rules that CSS grammar does in general: values are optional and re-orderable when possible, space-separated, and commas are used to separate repetitions only.

   For legacy reasons, `rgb()` and `hsl()` also supports an alternate syntax that separates all of its arguments with commas. Also for legacy reasons, the `rgba()` and `hsla()` functions exist using the same comma-based syntax."

**Comment by mathiasbynens** [comment]
Re: `"comma-free"|"comma-based"`, my concern was more about making it clear that one is legacy syntax, and the other is more modern and future-facing. If we're not gonna choose names that guide people towards using the modern option, we could just go with something like `"with-comma"|"without-comma"`. Either way the name doesn't matter too much and is not a blocker for prototyping this rule :)

I *love* the extended description you've put together. It concisely explains everything developers need to know. 👍

**Comment by jeddy3** [comment]
That's a good point. While working through the [degree notation issue](https://github.com/stylelint/stylelint/issues/4721#issuecomment-619522831) I realised that the intent of this rule is different to that one; this is about legacy vs modern syntax and that is about explicitness vs shorthand convenience.

We should use "modern" and "legacy" for the options as they don't obfuscate the intent of this rule.

> Either way the name doesn't matter too much and is not a blocker for prototyping this rule :)

I agree. 

I'll label the rule as ready to go.

- Name: `color-function-notation`
- Primary option: `"modern"|"legacy"`
- Secondary options: None
- Message: `Expected ${option} color-function notation`
- Description: "Specify modern or legacy notation for applicable color-functions."
- Section: ["Limit language features" -> "Color"](https://stylelint.io/user-guide/rules/list#color-1)
- Extended description: 
   - "Modern color-functions use a comma-free syntax because functions in CSS are used to group/name a syntax chunk. They should work by the same rules that CSS grammar does in general: values are optional and re-orderable when possible, space-separated, and commas are used to separate repetitions only."
   - "For legacy reasons, `rgb()` and `hsl()` also supports an alternate syntax that separates all of its arguments with commas. Also for legacy reasons, the `rgba()` and `hsla()` functions exist using the same comma-based syntax."

If anyone would like to [contribute](https://stylelint.io/contributing) this rule, please comment. There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

You'll likely want to use the [`selector-pseudo-element-colon-notation` rule](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-pseudo-element-colon-notation/) as a blueprint, but you'll want to use `postcss-value-parser` rather than `stylesearch` when analysing and fixing the color nodes. The [`color-named` rule](https://github.com/stylelint/stylelint/blob/master/lib/rules/color-named/index.js) has a good example of using the value parser for analysis.

**Comment by jeddy3** [comment]
Of the [commonly overlooked edge-cases](https://stylelint.io/developer-guide/rules#commonly-overlooked-edge-cases), you'll want to pay close attention to:

- variables
- CSS comments
- case sensitivity
- whitespace and punctuation

For example, the rule should autofix the following:

```css
a {
  color: HSLA(
    calc(var(--hue) + 30) /* comment */,
    28%, 50% /* comment */, 0.2);
}
```

As:

```css
a {
  color: HSL(
    calc(var(--hue) + 30) /* comment */
    28% 50% /* comment */ / 0.2);
}
```

Rules don't have to autofix every violation. If fixing these edge-cases, or even the entirety of the `"legacy"` option, is problematic, then they can be skipped.

**Comment by dsine-de** [comment]
Somehow `hsla()` is still allowed with the `modern` value (without commas between parameters), is this expected?
